src/shared/GitHub/GitHubHostProvider.cs

```diff
@@ -145,24 +145,42 @@ public async Task<ICredential> GetCredentialAsync(InputArguments input)
                     _context.Trace.WriteLine($"  {account}");
                 }
 
-                // If we have a WWW-Authenticate header then we can try and use any domain hint information
-                // to filter the list of accounts to only those that are valid for that domain.
-                // We only expect one challenge header to be returned, but if we're given more we just select the first.
-                GitHubAuthChallenge authChallenge = GitHubAuthChallenge.FromHeaders(input.WwwAuth).FirstOrDefault();
-                if (authChallenge is not null)
+                // Allow the user to disable account filtering until this feature stabilises.
+                // Default to enabled.
+                bool enableFiltering = !_context.Settings.TryGetSetting(
+                    GitHubConstants.EnvironmentVariables.AccountFiltering,
+                    Constants.GitConfiguration.Credential.SectionName,
+                    GitHubConstants.GitConfiguration.Credential.AccountFiltering,
+                    out string enableFilteringStr
+                ) || enableFilteringStr.ToBooleanyOrDefault(true);
```

Comment:
I'm a bit confused by this syntax. When will this `ToBooleanyOrDefault` only be invoked if `TryGetSetting` returns `true`?

Reply:
If we expand this to an `if` statement, it maybe becomes a little clearer:

```csharp
bool enableFiltering;
if (_context.Settings.TryGetSetting(
                    GitHubConstants.EnvironmentVariables.AccountFiltering,
                    Constants.GitConfiguration.Credential.SectionName,
                    GitHubConstants.GitConfiguration.Credential.AccountFiltering,
                    out string enableFilteringStr))
{
    bool? enableBooleany = enableFilteringStr.ToBooleany();
    if (enableBooleany.HasValue)
    {
        // Take the valid setting value as the answer
        enableFiltering = enableBooleany.Value;
    }
    else
    {
        // Default to `true` if the setting existed but wasn't a boolean
        enableFiltering = true;
    }
}
else
{
    // Default to `true` if no setting is present
    enableFiltering = true;
}
```

An abridged version could also be represented like:
```csharp
bool enable = !TryGetSetting(.., out string val) || (val.ToBool?() ?? true)
```

Enabled if no settings, OR if the value wasn't a bool. Otherwise, use the value.

Reply:
This is a common pattern we're using in the `Settings` class. Example: https://github.com/git-ecosystem/git-credential-manager/blob/0ae7f478ef4e7871a9c15086d1051654c5ba14d2/src/shared/Core/Settings.cs#L605-L606